Use `objdump` to determine whether a binary blob from a compressed archive should be handled by Carthage

blender opened the pull request:
Some frameworks don't have `Info.plists` and are being currently rejected as binary dependencies.

This PR addresses #2500 and also makes sure that .version files are written.

mdiep commented on Source/CarthageKit/Project.swift:
```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
+					.single()?
+					.value
+
+					return fileType != nil
```
I think this code ought to be split out into its own function.

mdiep replied:
I'd want to see tests for this as well.

blender replied:
ok

mdiep commented on Source/CarthageKit/VersionFile.swift:
```diff
@@ -49,7 +49,13 @@ struct VersionFile: Codable {
+		iOS: [CachedFramework]?,
+		watchOS: [CachedFramework]?,
+		tvOS: [CachedFramework]?
+	) {
```
Are there any functional changes in this file? I'm having trouble spotting anything with the reformatting.

blender replied:
No there are no functional changes.

jdhealy commented on Source/CarthageKit/Project.swift:
```diff
@@ -1289,14 +1290,78 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
+							|| $0.contains(String(format: "%x", MH_CIGAM_64))
+							|| $0.contains(String(format: "%x", MH_MAGIC_64))
+					}
+					.take(last: 1)
```
So, on the unstated assertion here (‘every `filetype` on a magic number containing line will be the same, so just grab any of them — as implemented: the last one’) — I said earlier that I’d post an implementation where a `reduce` makes this more explicit:

```swift
let acceptedFiletypesAsSpecifiedByMachHeader: Set<Int32> = [
	MH_OBJECT, // Carthage accepts static libraries
	MH_BUNDLE, // Bundles https://github.com/ResearchKit/ResearchKit/blob/1.3.0/ResearchKit/Info.plist#L15-L16
	MH_DYLIB, // or dynamic shared libraries
]

let øøø: Bool = task.launch(standardInput: nil)
	.ignoreTaskData()
	.map { String(data: $0, encoding: .utf8) ?? "" }
	.filter { !$0.isEmpty }
	.flatMap(.merge) { (output: String) -> SignalProducer<(String, String), NoError> in
		output.linesProducer.combinePrevious()
	}
	.filterMap {
		guard
			$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),
			[MH_CIGAM, MH_MAGIC, MH_CIGAM_64, MH_MAGIC_64]
				.lazy.map({ " " + String(format: "%x", $0) })
				.first(where: $0.1.hasPrefix) != nil
		else { return nil }

		let components = $0.1
			.components(separatedBy: CharacterSet.whitespaces)
			.filter { !$0.isEmpty }

		precondition(components.count == 8, "∆∆∆")

		return Int32(components[4])
	}
	.filter(acceptedFiletypesAsSpecifiedByMachHeader.contains)
	.reduce(into: Set<Int32>()) { $0.insert($1); return }
	.map { $0.count == 1 }
	.single()?
	.value ?? false
```

What do you think?

blender replied:
I like the reduce but I think that this line with is maybe too "whitespace" sensitive?

`$0.0.hasPrefix("      magic cputype cpusubtype  caps    filetype ncmds"),`

jdhealy replied:
Whitespace is fixed according to <https://llvm.org/viewvc/llvm-project/llvm/trunk/tools/llvm-objdump/MachODump.cpp?view=markup&pathrev=225383###see·line·2745> — IMO, we’d want a silent non-consideration if that output structure were to change (adding a field).

jdhealy replied:
I guess we could strip the whitespace from both the output line and our conditional and compare, that would still catch an added field.

blender replied:
I'm going to push a few changes according to @mdiep desire to have a separate function.

I'll try to put the reduce too

jdhealy commented:
Great work tracking this down (I should have said upfront) — massive thanks 👏

blender commented:
I have updated the PR:
- Introduced `MachHeader` strcture
- refactored out function to read mach headers in static method on `MachHeader`
- Added tests
- Remove checking for package type. It's superfluous give that the headers always work.

blender commented:
Any opinion on the refactored version?

mdiep commented:
The code looks fine to me, but I'm not totally confident whether this is a good idea. What do you think, @jdhealy?

blender commented on Source/CarthageKit/Project.swift:
```diff
@@ -1,6 +1,7 @@
 // swiftlint:disable file_length
 
 import Foundation
+import MachO.loader
```
No need. Good catch.

blender commented:
The alternative that I have discussed with @jdhealy is to read the filetype by hand. I think it would be more error prone than using existing tools that know how to read Mach-O files correctly.

jdhealy commented on Source/CarthageKit/Project.swift:
```diff
@@ -1282,23 +1283,33 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
+
+				return MachHeader.headers(forMachOFileAtUrl: executableURL)
+					.filter { MachHeader.carthageSupportedFileTypes.contains($0.fileType) }
+					.reduce(into: Set<UInt32>()) { $0.insert($1.fileType); return }
```
That `.insert` doesn’t return `Void` which necessitates either a `_ = ` prefix or a `return` suffix to match the type signature of the `into:` `reduce`.

jdhealy commented on Source/CarthageKit/MachHeader.swift:
```diff
@@ -0,0 +1,147 @@
+					let sizeofcmds = UInt32(strippedComponents[6], radix: 10),
+					let flags = UInt32(strippedComponents[7], radix: 16)
+					else {
+
```
Nit: I’d prefer this without this empty line.

blender replied:
Done

jdhealy commented on Source/CarthageKit/Project.swift:
```diff
@@ -1282,23 +1283,33 @@ internal func frameworksInDirectory(_ directoryURL: URL) -> SignalProducer<URL,
 		}.filter { url in
-
-			let packageType: PackageType? = Bundle(url: url)?.packageType
+			// Try reading the Info.plist of the Framework
```
I’m confident in the structure of the CLI output from `objdump` rarely changing (and if it does, we should be able to catch via open-sourcedness before it’s widely distributed), but a simple Plist read is probably faster than a process execution. If you agree, could we leave this the default and add a comment with the reasoning?

```
// For reasons of speed and the fact that CLI-output structures can change,
// first try the safer method of reading the ‘Info.plist’ from the Framework’s bundle.
```

blender replied:
Sure

blender commented on Source/CarthageKit/MachHeader.swift:
```diff
@@ -0,0 +1,147 @@
+	/// 	      The output is composed of one or more sets of lines like the following:
+	///
+	///			Mach header
+	///		          magic  cputype cpusubtype  caps    filetype ncmds sizeofcmds      flags
```
Something my new installation of Xcode is wrong and it's driving me crazy with spaces and tabs :D . Fixed.

blender commented:
I would like to rebase before merging to remove superfluous commits

blender commented:
Rebased. All Done.
